src/main/groovy/io/cify/framework/core/Capabilities.groovy

```diff
@@ -55,37 +74,78 @@ class Capabilities {
-                return new DesiredCapabilities(browser)
-            case DeviceCategory.ANDROID:
-                return new DesiredCapabilities(android)
-            case DeviceCategory.IOS:
-                return new DesiredCapabilities(ios)
+        Map result = [:]
+        try {
+            result = capabilityList.find {
+                (it.category as String).toUpperCase() == category.toString() && it.available
+            }.capabilities as LazyMap
+        } catch (NullPointerException) {
+            new CifyFrameworkException("No available capabilities for $category found")
         }
-        throw new CifyFrameworkException("Unsupported device category $category")
+        return new DesiredCapabilities(result)
+    }
+
+    /**
+     * Returns desired capabilities for given capabilityId
+     *
+     * @param capabilityId capabilities identifier
+     * @return DesiredCapabilities
+     */
+    DesiredCapabilities toDesiredCapabilities(String capabilityId) {
+        LOG.debug(MARKER, "Get desired capabilities for capabilityID $capabilityId")
+        Map result = [:]
+        try {
+            result = capabilityList.find { it.id == capabilityId && it.available }.capabilities as LazyMap
+        } catch (NullPointerException) {
+            new CifyFrameworkException("No available capabilities for $capabilityId found")
+        }
+        return new DesiredCapabilities(result)
     }
 
     /**
      * Adds to desired capabilities
+     *
      * @param category - Device category
      * @param key - key to add
      * @param value - value to add
      */
-    void addToDesiredCapabilities(DeviceCategory category, String key, def value) {
+    void addToDesiredCapabilities(DeviceCategory category, String key, String value) {
         LOG.debug(MARKER, "Add to desired capabilities for category $category")
-        switch (category) {
-            case DeviceCategory.BROWSER:
-                browser.put(key, value)
-                break
-            case DeviceCategory.ANDROID:
-                android.put(key, value)
-                break
-            case DeviceCategory.IOS:
-                ios.put(key, value)
-                break
-            default:
-                throw new CifyFrameworkException("Unsupported device category $category")
+        try {
+            capabilityList.findAll { (it.category as String).toUpperCase() == category.toString() }.each {
+                it.capabilities.put(key, value)
+            }
+        } catch (NullPointerException) {
+            throw new CifyFrameworkException("Could not add $key:$value. Capabilities for category $category not found.")
+        }
+    }
+
+    /**
+     * Sets available key of capability with given capabilityId to given value
+     *
+     * @param capabilityId id of the capability
+     *
+     * @param value value given to available key
+     */
+    void setAvailable(String capabilityId, boolean isAvailable) {
+        try {
```

Comment:
Is it make sense to check capabilityId against emptiness?

Reply:
I don't think so, as capabilityId is being autogenerated (if not provided) while json parsing